src/petclaw/solution.py

```diff
@@ -24,4 +25,89 @@ def get_write_func(self, file_format):
+
+    def _init_ds_state(self, state):
+        """
+        Returns a downsampled version of the given state object
+        """
+        import clawpack.petclaw as pyclaw
+
+        ds_domain = pyclaw.Domain([pyclaw.Dimension(self.domain.patch.lower_global[i],
+                                                    self.domain.patch.upper_global[i],
+                                                    self.domain.patch.num_cells_global[i]/self.downsampling_factors[i],
+                                                    self.domain.grid.dimensions[i].name)
+                                   for i in range(self.domain.num_dim)], proc_sizes=state.q_da.getProcSizes())
+        ds_state = pyclaw.State(ds_domain,state.num_eqn,state.num_aux)
+
+        return ds_state
+
+    def downsample(self, write_aux,write_p):
+        """
+        Returns a downsampled version of the solution by local averaging over the downsampling factors
+        """
+        for i  in range(len(self.states)):
+            state = self.states[i]
+            if i > 0:
+                self.ds_solution.states.append(self._init_ds_state(self.downsample_factors, state))
+            ds_state = self.ds_solution.states[i]
+
+            # Downsample q
+            if write_p:
+                ds_state.p = self._downsample_global_to_local(state.get_q_da_for_ds(np.max(self.downsampling_factors)),
+                                                             state.gqVec, state.num_eqn, ds_state.patch._da.getRanges())
+            else:
+                ds_state.q = self._downsample_global_to_local(state.get_q_da_for_ds(np.max(self.downsampling_factors)),
+                                                             state.gqVec, state.num_eqn, ds_state.patch._da.getRanges())
+            # Dowsample aux
+            if write_aux:
+                ds_state.aux = self._downsample_global_to_local(state.get_aux_da_for_ds(np.max(self.downsampling_factors)),
+                                                               state.gauxVec, state.num_aux, ds_state.patch._da.getRanges())
+
+        return self.ds_solution
+
+    def _downsample_global_to_local(self, da_for_ds, gVec, num_eqn, ds_domain_ranges):
+        """
+        Returns a locally averaged array and handles ranges mapping between the original & downsampled solution objects
+
+        Input:
+            - da_for_ds: the DA object of the original solution (q or aux) with stencil width adjusted as appropriate to
+              handle the boundaries of the array to be averaged because after domain decomposition, there are no guarantees that
+              the downsampling factors will evenly divide the the sub-domain in each direction
+            - gVec: The global vector associated with q or aux
+            - num_eqn: number of components of q or aux
+            - ds_domain_ranges: global ranges of the downsampled solution
+        """
+
+        from skimage.transform import downscale_local_mean
```

Comment:
Can we add some specifics as to what `downscale_local_mean` is doing?  For instance, what the stencil size is?  At the very least put in the doc-string that the function is from sci-kit image.